### PR TITLE
kernel: netfilter: fix build warnings and handle upstream module changes

### PR DESCRIPTION
**Pull request by guidosarducci**
**Review/Distribution List**:  @rmilecki , @xback , @blocktrron  (based on similar fixes and commits in the repo)
**Compile tested**:  mips32-be-malta and armvirt-64, master branch
**Run tested**:  mips32-be-malta and armvirt-64, master branch

**Description**:
Kernels 4.19 and 5.4 changed the organization of several netfilter modules, resulting in option changes or combined/moved functions. This PR updates the build configuration to reflect these changes:

1. Combine CONFIG_NF_NAT_MASQUERADE_{IPV4,IPV6} into CONFIG_NF_NAT_MASQUERADE.
2. Fold `nf_nat_redirect.o` function into `nf_nat.ko` module and make related config option boolean. 
3. Move the CT GRE code into the core `nf_conntrack.ko` module and change the CONFIG_NF_CT_PROTO_GRE option to boolean.

These updates should also address problems raised in FS#2476 and FS#2990.

Thanks for your review and feedback.

### CODE REVIEW

**Comment by guidosarducci** [comment]
Also pinging @Ansuel since he reported FS#2476 which this should address.

**Comment by guidosarducci** [comment]
@adrianschmutzler Aside from referencing related bugs, notifying their reporters, and pinging other committers of similar changes, is there anything else that can be done to encourage reviewing these fixes? Thanks.

File: include/netfilter.mk [hunk 1]
```diff
@@ -185,7 +185,7 @@ $(eval $(call nf_add,IPT_IPV6_EXTRA,CONFIG_IP6_NF_MATCH_RT, $(P_V6)ip6t_rt))
 
 # kernel only
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT, $(P_XT)nf_nat),))
-$(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT, $(P_XT)nf_nat_redirect),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT, $(P_XT)nf_nat_redirect, lt 4.18),))
```
Comment (yousong) [inline include/netfilter.mk:188]: Not that changes much, but to be precise it should be `le 4.18`

File: include/netfilter.mk [hunk 2]
```diff
@@ -223,7 +223,7 @@ $(eval $(call nf_add,NF_NATHELPER,CONFIG_NF_NAT_FTP, $(P_XT)nf_nat_ftp))
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CONNTRACK_BROADCAST, $(P_XT)nf_conntrack_broadcast))
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CONNTRACK_AMANDA, $(P_XT)nf_conntrack_amanda))
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_NAT_AMANDA, $(P_XT)nf_nat_amanda))
-$(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CT_PROTO_GRE, $(P_XT)nf_conntrack_proto_gre))
+$(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CT_PROTO_GRE, $(P_XT)nf_conntrack_proto_gre,lt 5.1))
```
Comment (yousong) [inline include/netfilter.mk:226]: ditto

**Review by yousong — APPROVED** [review]
Apart from the minor nitpicks, the series looks fairly good.  Thanks for the investigation and details explanation.  I plan to have this merged in the following days after the minors got fixed.

File: include/netfilter.mk (hunk 1 above)
Comment (guidosarducci) [reply include/netfilter.mk:188]: The module isn't present in kernel 4.18 and later, so 'lt 4.18' to build it is appropriate, no?

File: include/netfilter.mk (hunk 2 above)
Comment (guidosarducci) [reply include/netfilter.mk:226]: Likewise, the module disappeared when the option changed to boolean in kernel 5.1, so shouldn't 'lt 5.1' be correct?

File: include/netfilter.mk (hunk 1 above)
Comment (yousong) [reply include/netfilter.mk:188]: Lol, you are right!

File: include/netfilter.mk (hunk 2 above)
Comment (yousong) [reply include/netfilter.mk:226]: again, you are right ;)

**Comment by yousong** [comment]
Hi, I experimented the patch a bit and it generally works.  However I got a feeling that maybe something like the following change is better

```
diff --git a/include/netfilter.mk b/include/netfilter.mk
index 17a60d9ac6..d6e5297c83 100644
--- a/include/netfilter.mk
+++ b/include/netfilter.mk
@@ -186,6 +186,7 @@ $(eval $(call nf_add,IPT_IPV6_EXTRA,CONFIG_IP6_NF_MATCH_RT, $(P_V6)ip6t_rt))
 # kernel only
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT, $(P_XT)nf_nat),))
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT, $(P_XT)nf_nat_redirect, lt 4.18),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT=y,                     , ge 4.18),))
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_IPV4, $(P_V4)nf_nat_ipv4),))
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_MASQUERADE_IPV4, $(P_V4)nf_nat_masquerade_ipv4, lt 4.18),))
 
diff --git a/target/linux/generic/config-5.4 b/target/linux/generic/config-5.4
index 2f96cd86fa..026bc938c0 100644
--- a/target/linux/generic/config-5.4
+++ b/target/linux/generic/config-5.4
@@ -3673,7 +3673,6 @@ CONFIG_NF_CT_PROTO_GRE=y
 # CONFIG_NF_NAT_IPV6 is not set
 # CONFIG_NF_NAT_IRC is not set
 CONFIG_NF_NAT_MASQUERADE=y
-CONFIG_NF_NAT_REDIRECT=y
 # CONFIG_NF_NAT_NEEDED is not set
 # CONFIG_NF_NAT_PPTP is not set
 # CONFIG_NF_NAT_PROTO_GRE is not set
```

The point is that `CONFIG_NF_NAT_REDIRECT` on its own won't do.  It depends on `CONFIG_NETFILTER` and `CONFIG_NF_CONNTRACK`, both of them set at "package" level, as opposed to target level.

If this holds, we should also move other symbols like `CONFIG_NF_NAT_MASQUERADE` to `include/netfilter.mk`.  Ping @hauke and @jow- for comments & advice.

**Comment by guidosarducci** [comment]
> Hi, I experimented the patch a bit and it generally works.
>
Great, glad to hear it!
> However I got a feeling that maybe something like the following change is better
> ...
>  $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT, $(P_XT)nf_nat_redirect, lt 4.18),))
> +$(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT=y,                     , ge 4.18),))
>  ...
> 
How is it better? Please articulate more clearly. This seems more confusing to anyone trying to read and understand what the code is doing and why. Doesn't this also abuse/misuse the `nf_add` macro a little? (Does `CONFIG_NF_NAT_REDIRECT=y` actually work?)
> The point is that `CONFIG_NF_NAT_REDIRECT` on its own won't do.
>
But you just confirmed above that it works in general...
> It depends on `CONFIG_NETFILTER` and `CONFIG_NF_CONNTRACK`, both of them set at "package" level, as opposed to target level.
> 
The point is that `CONFIG_NF_NAT_REDIRECT=y` will be used if those other options are configured, but ignored otherwise. If your concern is more with a reliance on using settings at different levels, this is something we normally do, no? Lower level config settings override higher level defaults for example...
> If this holds, we should also move other symbols like `CONFIG_NF_NAT_MASQUERADE` to `include/netfilter.mk`.
>
What!? Why? The `CONFIG_NF_NAT_MASQUERADE` changes are just consolidating existing booleans and don't affect our configuration system. In fact, the previous OpenWRT changes with `CONFIG_NF_NAT_MASQUERADE_IPV4` and `CONFIG_NF_NAT_MASQUERADE_IPV6` mirror my current changes with `CONFIG_NF_NAT_REDIRECT`.
> Ping @hauke and @jow- for comments & advice.
>
Sure, that would be welcome. Please also see the following commit for some relevant precedent:
f891670704 ("kernel: netfilter: Adapt merge ipv4/ipv6 masquerade code")

**Comment by yousong** [comment]
> > Hi, I experimented the patch a bit and it generally works.
> 
> Great, glad to hear it!
> 
> > However I got a feeling that maybe something like the following change is better
> > ...
> > $(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT, $(P_XT)nf_nat_redirect, lt 4.18),))
> > +$(eval $(if $(NF_KMOD),$(call nf_add,NF_NAT,CONFIG_NF_NAT_REDIRECT=y,                     , ge 4.18),))
> > ...
> 
> How is it better? Please articulate more clearly. This seems more confusing to anyone trying to read and understand what the code is doing and why. Doesn't this also abuse/misuse the `nf_add` macro a little? (Does `CONFIG_NF_NAT_REDIRECT=y` actually work?)
> 

In my opinion it would be more clear instead of more confusing.  The relevant kconfig symbols are all in one place, not scattered somewhere in the tree with all others.  The conditional is there to make the version dependency clear.

Use of this way with `nf_add` should work.  Or at least `KCONFIG` makefile variable accepts such form as `CONFIG_SYMBOL=x`.  If `nf_add` has issue with this, we should probably make it work with it.

> > The point is that `CONFIG_NF_NAT_REDIRECT` on its own won't do.
> 
> But you just confirmed above that it works in general...
> 
> > It depends on `CONFIG_NETFILTER` and `CONFIG_NF_CONNTRACK`, both of them set at "package" level, as opposed to target level.
> 
> The point is that `CONFIG_NF_NAT_REDIRECT=y` will be used if those other options are configured, but ignored otherwise. If your concern is more with a reliance on using settings at different levels, this is something we normally do, no? Lower level config settings override higher level defaults for example...
> 
> > If this holds, we should also move other symbols like `CONFIG_NF_NAT_MASQUERADE` to `include/netfilter.mk`.
> 
> What!? Why? The `CONFIG_NF_NAT_MASQUERADE` changes are just consolidating existing booleans and don't affect our configuration system. In fact, the previous OpenWRT changes with `CONFIG_NF_NAT_MASQUERADE_IPV4` and `CONFIG_NF_NAT_MASQUERADE_IPV6` mirror my current changes with `CONFIG_NF_NAT_REDIRECT`.
> 
> > Ping @hauke and @jow- for comments & advice.
> 
> Sure, that would be welcome. Please also see the following commit for some relevant precedent:
> [f891670](https://github.com/openwrt/openwrt/commit/f8916707045170b2509c34f0d72f9bb4dae463f6) ("kernel: netfilter: Adapt merge ipv4/ipv6 masquerade code")

Yes, I am aware of this precedent by @hauke when bringing in 5.4 support.  The intent is that maybe this could be organized better.  Currently target kconfig settings contains XX=y that will resolve to "# XX is not set" unless some package was selected.  This is what seems confusing to me.  Settings should better be explicit and have direct effect, not implied by outside conditions.

**Comment by yousong** [comment]
Hi @guidosarducci , would you now like to give the above proposal a try?  There is no other comments on this yet but I still think it's promising.  I can also take it from here if that's okay with you, but it may take a while before I can get to it.

**Comment by guidosarducci** [comment]
@yousong The intent of this PR is simple: silence warnings due to kernel config changes over time, and do so in a way self-consistent with current OpenWRT practice. Now, I've gathered you don't agree with current practice, but I don't think this PR is a vehicle for changing that.

I can sympathize with the problems you outlined and the behaviour you'd like to see:
> ... relevant kconfig symbols are all in one place, not scattered somewhere in the tree with all others.
>
> The intent is that maybe this could be organized better. Currently target kconfig settings contains XX=y that will resolve to "# XX is not set" unless some package was selected. This is what seems confusing to me. Settings should better be explicit and have direct effect, not implied by outside conditions.

but these are much larger in scope than this PR, and TBH I haven't seen a proposal that analyzes the issues/consequences and outlines a solution.

If you'd like to pause or close this PR while you address the kernel warnings in the context of a larger reorganization you plan to investigate, that's OK with me. However, I would still like to see feedback from other members ( @hauke and @jow- ) since there's potentially a baby/bathwater situation where we simply trade one set of problems for another.

**Comment by guidosarducci** [comment]
@yousong  Have you made any progress on the changes you were thinking about? Any feedback from other members? Since I gather the 20.xx release is approaching, I will suggest pulling in this PR now and fixing the related bugs, while continuing work on your changes for a subsequent release.

**Comment by yousong** [comment]
> @yousong Have you made any progress on the changes you were thinking about? Any feedback from other members? Since I gather the 20.xx release is approaching, I will suggest pulling in this PR now and fixing the related bugs, while continuing work on your changes for a subsequent release. 

Sorry for the delay.  Inconvenient it may be, I don't see the issue urgent enough to qualify an immediate merge of this pull request in its current form, especially when there is a possible better solution (not too high the fruit maybe).

Guess I will have to work on this in the following days.  That said, it's also okay with me should this be handled quicker by others, in case the 20.xx release schedule is so close, or for whatever other reasons.

**Comment by yousong** [comment]
Hi @guidosarducci , I just pushed a series [1] to openwrt master.  When you find time available, please check if they fix the issues also reported here

Thank you.

 [1] http://lists.openwrt.org/pipermail/openwrt-devel/2020-October/031823.html

**Comment by yousong** [comment]
I am closing this pull request now that the issues reported here should be fixed by another set of patches mentioned above.

@guidosarducci , Thank you for the heads up and all the time and efforts pushing this

**Pull request closed without merging**